sympy/core/basic.py

```diff
@@ -1258,6 +1258,33 @@ def has(self, *patterns):
         """
         return self._has(iterargs, *patterns)
 
+    def has_xfree(self, s: set[Basic]):
+        """return True if self has any of the patterns in s as a
+        free argument, else False. This is like `Basic.has_free`
+        but this will only report exact argument matches.
+
+        Examples
+        ========
+
+        >>> from sympy import Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> f(x).has_xfree({f})
+        False
+        >>> f(x).has_xfree({f(x)})
+        True
+        >>> f(x + 1).has_xfree({x})
+        True
+        >>> f(x + 1).has_xfree({x + 1})
+        True
+        >>> f(x + y + 1).has_xfree({x + 1})
+        False
+        """
+        # protect O(1) containment check by requiring:
+        if not type(s) in (dict, set):
+            raise ValueError('expecting set or dict argument')
```

Comment:
I guess TypeError is more appropriate. The type hint in the signature suggests that only `set` is allowed. I suggest only allowing `set` here. I don't see a strong benefit in anything else.

Reply:
I put the dict there because you indicated that dict or set would have O(1) lookup. And there is no reason to force the user to convert dict to set if it can be used directly. I can make up a case where someone would want to pass a dict, but I'm not sure *I* would do so in practice. But I have appreciated the flexibility of Python to just "do the right thing" if no rules are broken...and to enforce as few rules as necessary.